src/deepsparse/loggers/function_logger.py

```diff
@@ -0,0 +1,111 @@
+from deepsparse.loggers.metric_functions import apply_function
+
+
+__all__ = ["FunctionLogger"]
+
+
+class FunctionLogger(BaseLogger):
+    """
+    DeepSparse logger that applies functions to raw log values
+    according to the specified config file
+
+    :param loggers: A DeepSparse Logger object
+    :param config: A configuration dictionary that specifies the mapping between
+        the target name and the functions to be applied to raw log values
+        e.g.
+        ```
+        {"pipeline_inputs":
+            [{"function": "identity",
+              "frequency": 3},
+
+            [{"function": "identity",
+              "frequency": 5}],
+
+         "pipeline_outputs":
+            [{"function": "identity",
+            "frequency": 4}]
+        }
+    """
+
+    def __init__(
+        self,
+        logger: BaseLogger,
+        config: Dict[str, List[Dict[str, Any]]],
+    ):
+
+        self.logger = logger
+        self.config = self._add_frequency_counter(config)
+
+    def log(self, identifier: str, value: Any, category: MetricCategories):
+        """
+        :param identifier: The identifier of the log
+             By default should consist of at least one string (name of the pipeline):
+                e.g. identifier = "pipeline_name"
+            If identifier is to be more complex, optional strings are to be concatenated
+            using dot as a separator:
+                e.g. identifier = "pipeline_name.some_argument_1.some_argument_2"
+        :param value: The data structure that the logger is logging
+        :param category: The metric category that the log belongs to
+        """
+        if category == MetricCategories.DATA:
+            pipeline_name, *target_name = identifier.split(".")
+
+            list_functions = self.config.get(".".join(target_name))
+            if list_functions is None:
+                return
+
+            for function_dict in list_functions:  # iterate over all available functions
+                if function_dict["frequency_counter"] % function_dict["frequency"] == 0:
+                    # if count matches the frequency, log
+                    # otherwise, pass
+                    self._log_data(identifier, value, category, function_dict)
+                # increment the counter
+                function_dict["frequency_counter"] += 1
+
+        else:
+            self.logger.log(identifier=identifier, value=value, category=category)
+
+    def _log_data(
+        self,
+        identifier: str,
+        value: Any,
+        category: MetricCategories,
+        function_dict: Dict[str, Any],
+    ):
+        # fetch the string signature & apply function
+        # specified by the signature to the value
+        function_signature = function_dict.get("function")
+        value = apply_function(value=value, function_signature=function_signature)
+        self.logger.log(
+            identifier=".".join([identifier, function_signature]),
```

Comment:
might need to think about this a bit more since an identifier could have multiple functions with the same name, but different kwargs

Reply:
we are not going to support kwargs in yaml - going for simple builtins with one purpose or dynamic function imports. its easier for us to document and easier for users to understand what to do.

if a user wants to do something not supported by builtins they should add custom functions (which will be very easy) rather than exposing kwargs in yaml.

Reply:
I share @corey-nm sentiment, will release my architectural proposal today that encapsulates this thought.

Reply:
@bfineran I continue the thread in our slack group.